### DISCUSSION TITLE
DYI - Strommessung über Hall - Sensor -> SOC ermitteln

### DISCUSSION THREAD

**Discussion by Ralf-64** [hoylabs/OpenDTU-OnBattery#1142] [Ideas]
Hallo zusammen,
ein nettes Feature für OpenDtu on Battery, wäre in meinen Augen eine Onboard Strommessung des Akku Lade/Entladesromes. Auch zur Ermittlung des SOC / DOD.
In diversen Shops werden Sensoren des Typs zb. WCS1600 (100A) oder WCS1800 (35A) zu günstigen Preisen.
Einen solchen Sensor könnte man mit dem Onboard ADC des ESP einlesen.
Hat jemand in der Richtung schon was gemacht ??
Besteht in der Gruppe interesse an einem solchen Projekt ??
Das anpassen/ anbinden des Sensors an den ADC würde gerne unterstützen (Elektroniker).
LG Ralf

**Comment by swingstate** [comment]
Ich würde das interessant finden. Soll das einen Shunt ersetzen bzw. integrieren?

**Comment by SW-Niko** [comment]
Hallo @Ralf-64,
technisch sehe ich bei deinem Vorhaben keine großen Hürden. 
Ich frage mich nur was du mit dem Batteriestrom anfangen willst?  Vom SOC und DOD kommen wir langsam wegen fehlender Genauigkeit weg. Und ich habe bedenken, das wir mit einem überschaubaren Aufwand ein besseres Ergebnis, als namhafte Hersteller, auf die Beine stellen können.
Bitte nicht falsch verstehen. Ich will dir dein Vorhaben nicht madig reden.

**Comment by spcqike** [comment]
Was ist denn der Vorteil eines Hallsensors? Ist ein Shunt nicht flexibler und genauer?

Man könnte sich auch einen DIY Smartshunt basteln oder sowas in die openDTU integrieren. vorausgesetzt, man hat freie Pins.

https://github.com/kscholty/SmartShuntINA226

**Reply by SW-Niko** [reply]
Der Hall-Sensor umgeht schon mal das Masse-Problem. Damit kannst du mühelos auf der +DC oder auf der -DC Leitung messen. Das kannst du mit einem Shunt auch, aber dann mit mehr Aufwand.
Die Hall-Sensoren haben technologiebedingt ein Problem mit kleinen Strömen. Der Messfehler steigt dann deutlich an.

**Comment by Ralf-64** [comment]
Hallo zusammen.
Zunächst vielen Dank für das Interesse und die konstruktiven Vorschläge...

Habe ein paar Messungen an 2 vorhandenen WCS1600 (35A Ausführung) durchgeführt. Was soll ich sagen: Bin begeistert.
Die 0-Spannung liegt auf ca. halber Versorgungsspannung, 2,588 V bei dieser Platine (Platine B- 2,566V). 
Die Ausgangsspannung folgt, auch bei kleinen Änderungen, sauber der Stromänderung. Das hätte ich so nicht erwartet.
Falls jemand die Messreihen interessiert, gerne melden.

Bei "winson.com.tw" liegen ausführliche Datenblätter für alle WCS1xxx Sensoren zum download bereit (PDF)

Die Idee ist nun die Messwerte an einen Spannung nach I2C Wandler (ADS1115, die kleine Platine zwischen den Meßgeräten) auszugeben.
Diesen Datensatz könnte man dann in OpenOnBattery einlesen oder mit einem ESP als MQTT ausgeben.
Werde den Meßausbau die Tage um den ADS1115 erweitern und mit einem Arduino Uno auszuwerten.
Mit den ESP´s bin ich nicht so "per Du" ;-) 

@spcqike: Der große Vorteil des Hallo Sensors ist (wie SW-Niko schon sagte) die galanvische Trennung des Messignales.
Auch ist es einigermaßen umständlich eine 16² Leitung aufzutrennen, Kableschuhe aufzucrimpen, 8mm Schrauben auf einer 
isolierten Grundplatte zu befestigen und mit der Shuntplatine zu verlöten. 
Aber machbar.
Bin einer Shunt Lösung nicht prinzipiell abgeneigt. Die Überlegung geht aber dann eher in Richtung -> Spannungabfall über dem Akku Kabel zu messen.

@SW-Nico: Ja, es geht mir um den SOC / DOD. Das Ergebnis wird wohl nicht genauer als von den namhaften Herstellern.
Mir ist allerdings kein Einzelgerät auf dem Markt bekannt welches mit Hall Sensoren arbeitet und sich in OpenOnBatterie einbinden läßt oder MQQT sendet.
Einzige (kaufliche) Option ist wohl der Victron Shunt mit VE-Anschluß. Kennst du noch andere (standallone)Geräte ? 
Und: Du machst mich neugierig: Welche Alternativen zur SOC Ermittlung gibt es noch ? Spannungsüberwachung fällt bei LiFePo4 Zellen wohl aus...

LG Ralf

Der aktuelle Testaufbau:
![Testaufbau_Hall1](https://github.com/user-attachments/assets/d26b184a-f380-4414-bf39-f6bae6b470a8)

**Reply by SW-Niko** [reply]
Hallo @Ralf-64,
nein ich kenne leider keine anderen (standalone) Geräte ?
Aber ich habe schon vor Jahren Hall-Sensoren in einigen Entwicklungen eingesetzt. 
Hast du dir schon mal überlegt was für Anforderungen du stellt und welche Messgenauigkeit du erreichen must?
Ich mach mal ein Beispiel:
Der Hall-Sensor hat laut Datenblatt bis zu 2% Fehler. Wenn ich jetzt alle anderen Fehlerquellen ignoriere und 10A für 5h in eine Batterie mit 24V lade dann sind das E=24V x 10A x 5h = 1200Wh und 2 % davon sind 24Wh. 
Das ist als Fehler sicherlich zu verkraften. Aber wenn du das 5 Tage lang machst mit laden und entladen dann kannst du schon bei 240Wh als Fehler landen. In Realität wird es sicherlich weniger sein weil der Fehler sich teilweise wieder ausgleicht.
Im Endeffekt landest du bei der Frage: "In welchen Abständen muss ich eine Vollladung erreichen damit die Messfehler wieder entfernt werden". 
Das ist genau der Punkt mit dem auch die namhaften Hersteller ein Problem haben. Die Messgenauigkeit von Strom, Spannung und Zeit wirken sich gravierend auf die SoC aus wenn nicht regelmäßig eine Vollladung erreicht wird.
Darum messen die vermutlich auch über einen Shunt, weil damit preisgünstig Genauigkeiten von 0.1% kein Problem sind.
